### ISSUE TITLE
[expr.const] lvalue-to-rvalue conversion is not diagnosed to disqualify constant expressions for empty classes

### ISSUE THREAD

**Issue by Eisenwave** [llvm/llvm-project#83712]
https://godbolt.org/z/KnPx1ejbb
```cpp
struct S {};

constexpr int f(S s) {
    constexpr S _ = s;
    return 0;
}

S s;
constexpr int _ = f(s);
```
This code compiles, but should produce errors for both `constexpr` initializations.

According to [[expr.const] p5.9](http://eel.is/c++draft/expr.const#5.9), the initializers of `_` cannot be constant expression because they contain lvalue-to-rvalue conversion of `s`, whose lifetime did not begin within the evaluation of the constant expression, and which is not usable in constant expressions.

It shouldn't matter whether `S` is an empty class or has members; the lvalue-to-rvalue conversion in itself disqualifies expressions from being constant expressions.

**Comment by Eisenwave** [comment]
Corresponding GCC bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=114219 (confirmed, regression).

**Comment by pinskia** [comment]
But note from https://gcc.gnu.org/bugzilla/show_bug.cgi?id=91953#c3:
```
I think it should be accepted, since S is empty, there's no actual lvalue-to-rvalue conversion involved.

It works with
using S = decltype(nullptr);

and this works too since GCC 7:

struct A { } a;
constexpr int f (A a) { return 42; }
constexpr int i = f(a);
```

**Comment by Eisenwave** [comment]
I don't see how lvalue-to-rvalue conversion would be bypassed here.

[[conv.lval], lvalue-to-rvalue conversion](https://eel.is/c++draft/conv.lval#:conversion,lvalue-to-rvalue) has no special provision for empty classes.

[[dcl.init.general] p16.9](https://eel.is/c++draft/dcl.init.general#16.9) would necessitate lvalue-to-rvalue conversion because the initializer has to be converted to a prvalue. I couldn't find any special rule for empty classes.

**Comment by killerbee13** [comment]
[dcl.init.general] p16.6 has precedence over p16.9. Specifically, p16.6.2 because the type is the same. The overload resolution selects the implicitly-declared copy constructor, so `S::S(const S&)` is called, which is constexpr-suitable, and no lvalue-to-rvalue conversion is applied to convert to the argument type because it is a reference. Additionally, no lvalue-to-rvalue conversion occurs within the implicitly-defined body of that function, because there are no members to initialize.

There is no special rule for empty classes, it's just that the general rules for classes mean that the copy constructor has the form:
```cpp
constexpr S::S(const S&) noexcept { }
```
and that function would be legal to call in this case as you can see in this example:
https://godbolt.org/z/a8j5599dG

**Comment by Sirraide** [comment]
Yeah, from what I can tell, this never actually performs lvalue-to-rvalue conversion or accesses the ‘value’ of the object (if there even is one; it’s empty, after all); I also can’t find anything in [[expr.const]](https://eel.is/c++draft/expr.const#5) that would indicate that binding a reference to a non-`constexpr` lvalue is a problem, and the GCC devs seem to agree with that from what I can gather from the GCC bug linked above.

@tbaederr, @shafik I think Clang is correct here, what do you think?

**Comment by zygoloid** [comment]
Clang is correct here. In C++, the lvalue-to-rvalue conversion is almost never applied to class types (and I tend to think the places where it is, eg [expr.call]/11, are wording bugs...)

**Comment by Sirraide** [comment]
I figured as much, but thanks for confirming it. We can close this one then.

**Comment by shafik** [comment]
> Clang is correct here. In C++, the lvalue-to-rvalue conversion is almost never applied to class types (and I tend to think the places where it is, eg [expr.call]/11, are wording bugs...)

Should there be a core issue? Do you have a specific case in mind?

**Comment by t3nsor** [comment]
I don't think it's really a defect that we have (rare) lvalue-to-rvalue conversions for class types, since the behaviour of such lvalue-to-rvalue conversions is clearly defined by [conv.lval]/3.2. It's more of a quirk.

**Comment by zygoloid** [comment]
> I don't think it's really a defect that we have (rare) lvalue-to-rvalue conversions for class types, since the behaviour of such lvalue-to-rvalue conversions is clearly defined by [conv.lval]/3.2. It's more of a quirk.

Yeah, maybe you're right, but it's ... weird ... that almost all class copy-initialization from the same type is not considered an lvalue-to-rvalue conversion, but the extremely rare cases of a discarded-value expression of class type and passing a class glvalue through an ellipsis both perform lvalue-to-rvalue conversions, and special rules apply to those two cases that don't apply in general. It just seems like an unneeded rule change in a corner case, that's going to invite implementation bugs.

@shafik A specific case would look like this:

```c++
constexpr int f(auto) { return 0; }
constexpr int g(...) { return 0; }

struct Empty {};
Empty e;

// OK, of course
constexpr int a = f(Empty{});
constexpr int b = g(Empty{});

// OK: copy constructor doesn't access `e`
constexpr int c = f(e);

// Error, lvalue-to-rvalue conversion applied to `e`, whose value is unknown
// Non-constant by [expr.const]/5.9.
constexpr int d = g(e);
```

MSVC gets this right; GCC, Clang, and EDG all incorrectly accept `d`.

The last time I filed an issue on these weird class lvalue-to-rvalue conversions (CWG1741, specifically for the impact of lvalue-to-rvalue conversions on odr-use), we eventually ended up making lvalue-to-rvalue conversions on class types through `...` "not count" as an lvalue-to-rvalue conversion in that context -- see CWG2083. Maybe CWG would want to do the same thing here, and say that [expr.const]/5.9 doesn't apply to class glvalues.
